File: mobile/android/FUN/app/src/main/java/com/fun/app/core/playlist/PlaylistManager.kt
/**
 * FUN App - Playlist Manager
 * Handles episode queues and navigation for different viewing modes
 */

package com.fun.app.core.playlist

import com.fun.app.core.storage.WatchHistoryManager
import com.fun.app.data.models.*

object PlaylistManager {
    
    // MARK: - Create Discover Playlist
    
    fun createDiscoverPlaylist(series: List<Series>, watchHistory: WatchHistoryManager): List<Episode> {
        val playlist = mutableListOf<Episode>()
        
        for (seriesItem in series) {
            // Determine which episode 1 to show
            var targetEpisode: Episode? = null
            
            if (!seriesItem.seasons.isNullOrEmpty()) {
                // Multi-season series
                val progress = watchHistory.getSeriesProgress(seriesItem.id)
                
                if (progress != null && progress.completedSeasons.isNotEmpty()) {
                    // User has completed seasons - show next unwatched season's E1
                    val nextSeasonNumber = progress.completedSeasons.maxOrNull()!! + 1
                    val nextSeason = seriesItem.seasons.find { it.seasonNumber == nextSeasonNumber }
                    if (nextSeason != null && nextSeason.episodes.isNotEmpty()) {
                        targetEpisode = nextSeason.episodes.first()
                    }
                } else {
                    // Show Season 1, Episode 1
                    val firstSeason = seriesItem.seasons.find { it.seasonNumber == 1 }
                    if (firstSeason != null && firstSeason.episodes.isNotEmpty()) {
                        targetEpisode = firstSeason.episodes.first()
                    }
                }
            } else if (!seriesItem.episodes.isNullOrEmpty()) {
                // Single season series
                targetEpisode = seriesItem.episodes.first()
            }
            
            if (targetEpisode != null) {
                playlist.add(targetEpisode)
            }
        }
        
        // Shuffle for discovery
        return playlist.shuffled()
    }
    
    // MARK: - Create Binge Playlist
    
    fun createBingePlaylist(series: Series, startEpisodeId: String): List<Episode> {
        val allEpisodes = series.allEpisodes
        val startIndex = allEpisodes.indexOfFirst { it.id == startEpisodeId }
        
        return if (startIndex >= 0) {
            allEpisodes.drop(startIndex)
        } else {
            allEpisodes
        }
    }
    
    // MARK: - Create Series Playlist
    
    fun createSeriesPlaylist(series: Series, startEpisodeId: String, seasonNumber: Int? = null): List<Episode> {
        val allEpisodes = if (seasonNumber != null && series.seasons != null) {
            // Specific season only
            val season = series.seasons.find { it.seasonNumber == seasonNumber }
            season?.episodes?.sortedBy { it.episodeNumber } ?: emptyList()
        } else {
            // All episodes
            series.allEpisodes
        }
        
        val startIndex = allEpisodes.indexOfFirst { it.id == startEpisodeId }
        
        return if (startIndex >= 0) {
            allEpisodes.drop(startIndex)
        } else {
            allEpisodes
        }
    }
    
    // MARK: - Get Next Unwatched Episode
    
    fun getNextUnwatchedEpisode(series: Series, watchHistory: WatchHistoryManager): Episode? {
        val seriesHistory = watchHistory.getSeriesHistory(series.id)
        val watchedIds = seriesHistory.filter { it.completed }.map { it.episodeId }.toSet()
        
        return series.allEpisodes.firstOrNull { !watchedIds.contains(it.id) }
    }
}

File: mobile/ios/FUN/FUN/Core/Playlist/PlaylistManager.swift
/**
 * FUN App - Playlist Manager
 * Handles episode queues and navigation for different viewing modes
 */

import Foundation

class PlaylistManager {
    
    // MARK: - Create Discover Playlist
    
    static func createDiscoverPlaylist(series: [Series]) -> [Episode] {
        var playlist: [Episode] = []
        let watchHistory = WatchHistoryManager.shared
        
        for seriesItem in series {
            // Determine which episode 1 to show
            var targetEpisode: Episode?
            
            if let seasons = seriesItem.seasons, !seasons.isEmpty {
                // Multi-season series
                let progress = watchHistory.getSeriesProgress(seriesId: seriesItem.id)
                
                if let progress = progress, !progress.completedSeasons.isEmpty {
                    // User has completed seasons - show next unwatched season's E1
                    let nextSeasonNumber = progress.completedSeasons.max()! + 1
                    if let nextSeason = seasons.first(where: { $0.seasonNumber == nextSeasonNumber }),
                       let firstEp = nextSeason.episodes.first {
                        targetEpisode = firstEp
                    }
                } else {
                    // Show Season 1, Episode 1
                    if let firstSeason = seasons.first(where: { $0.seasonNumber == 1 }),
                       let firstEp = firstSeason.episodes.first {
                        targetEpisode = firstEp
                    }
                }
            } else if let episodes = seriesItem.episodes, let firstEp = episodes.first {
                // Single season series
                targetEpisode = firstEp
            }
            
            if let episode = targetEpisode {
                playlist.append(episode)
            }
        }
        
        // Shuffle for discovery
        return playlist.shuffled()
    }
    
    // MARK: - Create Binge Playlist
    
    static func createBingePlaylist(series: Series, startEpisodeId: String) -> [Episode] {
        let allEpisodes = series.allEpisodes
        
        guard let startIndex = allEpisodes.firstIndex(where: { $0.id == startEpisodeId }) else {
            return allEpisodes
        }
        
        // Return episodes starting from the start episode
        return Array(allEpisodes[startIndex...])
    }
    
    // MARK: - Create Series Playlist
    
    static func createSeriesPlaylist(series: Series, startEpisodeId: String, seasonNumber: Int? = nil) -> [Episode] {
        var allEpisodes: [Episode] = []
        
        if let seasonNum = seasonNumber, let seasons = series.seasons {
            // Specific season only
            if let season = seasons.first(where: { $0.seasonNumber == seasonNum }) {
                allEpisodes = season.episodes.sorted { $0.episodeNumber < $1.episodeNumber }
            }
        } else {
            // All episodes
            allEpisodes = series.allEpisodes
        }
        
        guard let startIndex = allEpisodes.firstIndex(where: { $0.id == startEpisodeId }) else {
            return allEpisodes
        }
        
        return Array(allEpisodes[startIndex...])
    }
    
    // MARK: - Get Next Unwatched Episode
    
    static func getNextUnwatchedEpisode(series: Series) -> Episode? {
        let watchHistory = WatchHistoryManager.shared
        let seriesHistory = watchHistory.getSeriesHistory(seriesId: series.id)
        let watchedIds = Set(seriesHistory.filter { $0.completed }.map { $0.episodeId })
        
        let allEpisodes = series.allEpisodes
        return allEpisodes.first { !watchedIds.contains($0.id) }
    }
    
    // MARK: - Prefetch Next Episode
    
    static func prefetchNextEpisode(_ episode: Episode?) {
        guard let episode = episode,
              let urlString = episode.videoUrl,
              let url = URL(string: urlString) else {
            return
        }
        
        // Prefetch video
        URLSession.shared.dataTask(with: url) { _, _, _ in
            // Prefetch initiated
        }.resume()
        
        // Prefetch thumbnail
        if let thumbnailURL = URL(string: episode.thumbnailUrl) {
            URLSession.shared.dataTask(with: thumbnailURL) { _, _, _ in
                // Prefetch initiated
            }.resume()
        }
    }
}
